Allow username, password and key from env/secret

zrthxn opened the pull request:
Closes: https://github.com/FNNDSC/ChRIS_store/issues/20

## Description
Added shell code to allow Swift username, password and key to be set via environment variables or docker secrets.

### Behavior of the variables being set
This is how the decision to choose where to take credentials from is taken.
1. If environment variables are given via the -e option, use these variables first.
2. Else if `credentials.env` file exists, use file to set environment variables.
3. If docker secret `swift-credentials` is set and the file exists, take precedence over environment variables.

[UPDATED]
There are three environment variables/secrets that can be sent
- SWIFT_USERNAME
- SWIFT_KEY

These can be added to the run command like so
```bash
$ docker run -d --name swift-onlyone -p 12345:8080 -e SWIFT_USERNAME=myuser:mypass -e SWIFT_KEY=mykey -v swift_storage:/srv -t fnndsc/docker-swift-onlyone 
```

zrthxn committed:
Allow username, password and key from env/secret

zrthxn commented:
@jbernal0019 @jennydaman Related to the hardcoded secrets issue in Swift storage service.

zrthxn committed:
Updated README

zrthxn committed:
Fix readme incorrect command

jennydaman commented on files/startmain.sh:
```diff
@@ -57,14 +61,42 @@ if [ ! -z "${SWIFT_STORAGE_URL_SCHEME}" ]; then
+	echo "Using default credentials file: /etc/swift/credentials.env ..."
+	cp /etc/swift/credentials.env /etc/swift/credentials.temp
+	cat /etc/swift/credentials.temp | while read line; do
+		sed -i -e "s/$line/export $line/g" /etc/swift/credentials.temp
```
Are you trying to prepend `export` to every line of the file?

```shll
sed -i -e 's/^/export /g' /etc/swift/credentials.temp
```

zrthxn replied:
Yes but I thought with the way I did it, you'd be able to control which lines to add `export` to by checking `$line`, like for example to check whether or not the line has an acceptable variable name.
Although I suppose its not that big of an issue if there are a few invalid env variables?

jennydaman commented on files/startmain.sh:
```diff
@@ -57,14 +61,42 @@ if [ ! -z "${SWIFT_STORAGE_URL_SCHEME}" ]; then
+	echo "Using environment credentials..."
+elif [ -e /etc/swift/credentials.env ]; then
+	echo "Using default credentials file: /etc/swift/credentials.env ..."
+	cp /etc/swift/credentials.env /etc/swift/credentials.temp
```
I recommend using `mktemp`

zrthxn committed:
Simplify credentials temp file

jbernal0019 commented:
This looks like a great contribution @zrthxn! Could we instead of having three env variables just have these two?:

1. SWIFT_USERNAME
2. SWIFT_KEY

For instance `SWIFT_USERNAME=chris:chris1234` and `SWIFT_KEY=testing`.  Notice that both the user and user's password are defined in  `SWIFT_USERNAME` separated by a colon. That way we won't need to change any of our current deployments and development files.

zrthxn commented:
@jbernal0019 Yes of course, that's what I had thought of initially. It needed some pattern matching so I didn't like that approach. But I'll try it out anyway.

zrthxn committed:
MAINTAINER field is deprecated

zrthxn committed:
Reduce number of variables

zrthxn committed:
Update readme according to variables

zrthxn committed:
Fix readme: silly mistake

zrthxn committed:
Deafult username value reset

jennydaman commented on files/startmain.sh:
```diff
@@ -57,14 +60,41 @@ if [ ! -z "${SWIFT_STORAGE_URL_SCHEME}" ]; then
 fi
 
+# If secrets given, take precedence over variables
+if [ -e /run/secret/swift-credentials ]; then
```
I would recommend allowing for the user to either:

- specify credentials via environment variables `SWIFT_USERNAME`
- specify credentials by giving a file path via environment variable `CREDENTIALS_FILE` (takes precedent)

The code would look something like this:

```bash
SWIFT_USERNAME="${SWIFT_USERNAME:-chris:chris1234}"
if [ -v CREDENTIALS_FILE ]; then
  source $CREDENTIALS_FILE
fi
export SWIFT_USERNAME
```

Also note: instead of using `sed` to export everything in the given `.env`, it's safer (and faster) to selectively export explicitly what we want: only `SWIFT_USERNAME` and `SWIFT_KEY`
... that is, if we need to be exporting these variables anyway. It seems like they're not being used by subprocesses and only as local variables, then written into `/etc/swift/dispersion.conf`?

Instead of using `sed` again on `/etc/swift/dispersion.conf` I think a heredoc would be easier (and faster)

```bash
cat > /etc/swift/dispersion.conf << EOF
[dispersion]
auth_url = http://127.0.0.1:8080/auth/v1.0
auth_user = $SWIFT_USERNAME
auth_key = $SWIFT_KEY
endpoint_type = internalURL
EOF
```

zrthxn replied:
So instead of prepending "export" and then source-ing that file, like here
https://github.com/FNNDSC/docker-swift-onlyone/blob/39943ff5fadab12e61d978def835deb4032ff29b/files/startmain.sh#L80-L82

I can just extract the variable values from the file and put those into the local variables. Then continue using `sed` to modify the `/etc/swift/proxy-server.conf` and `/etc/swift/dispersion.conf` files, since that allows other values to remain unchanged from what the user set them to, like the `auth_url`.

jennydaman replied:
I recommend using a heredoc instead of `awk`/`sed`/`cut` because it's more legible and does not use child processes.

zrthxn replied:
@jennydaman How would I go about keeping the rest of the file unchanged/as it was (modified or not), with a heredoc?

jennydaman replied:
Maybe distasteful, but put the entire config file as a heredoc inside the script.

zrthxn replied:
Even though that'll make the conf files sort of ineffective?

jennydaman replied:
What do you mean by ineffective?
Do you mean, `git rm` them?

zrthxn replied:
No I mean if the heredoc is setting the content of the file, then whatever changes you make to the config before building the image won't work because the script will modify the file and remove them.

jennydaman replied:
Do you think we should support having the image be customizable by adding your own files, or should we only support customization via the environment variables we've chosen to expose?

zrthxn replied:
I think it should be customisable. This isn't an app that we're building, it's essentially just a config for setting up Swift. So I think it should allow maximum configuration of the actual application underneath, otherwise it sort of misses the point, doesn't it?

jennydaman replied:
Earlier today we've discussed senarios where less customization is preferable.

This is a somewhat opinionated configuration of swift, intended for development purposes. Hence I think less (and/or easier) code and fewer decisions is advantageous.

zrthxn replied:
Alright then I'll use the the suggested method.

zrthxn committed:
Improve variable extraction from credentials files

jennydaman commented on files/startmain.sh:
```diff
@@ -57,14 +60,40 @@ if [ ! -z "${SWIFT_STORAGE_URL_SCHEME}" ]; then
+	echo "Using environment credentials..."
+elif [ -e /etc/swift/credentials.env ]; then
+	echo "Using default credentials file: /etc/swift/credentials.env ..."
+	SWIFT_USERNAME = awk '/SWIFT_USERNAME/' /etc/swift/credentials.env | cut -f2 -d=
```
A PCRE look-behind

```
grep -m 1 -Po '(?<=SWIFT_USERNAME).+'
```

zrthxn replied:
This regex is not working right. It doesn't get any results.

![Screenshot (66)](https://user-images.githubusercontent.com/35369637/117434077-da027680-af49-11eb-9794-6bd4fa03e4d5.png)

I found that this works, but you have to explicitly specify the class of characters to look-behind from.

```bash
grep -m 1 -Po '(?<=SWIFT_USERNAME.)[A-Za-z0-9\:]*'
```

![Screenshot (69)](https://user-images.githubusercontent.com/35369637/117434186-fe5e5300-af49-11eb-9b73-d54abe57fe06.png)

jennydaman replied:
What is the output of `grep --version`?

Have you tried using a container?

```bash
$ cat example.env | docker run --rm -i ubuntu:18.04 grep -m 1 -Po '(?<=SWIFT_USERNAME).+
=chris:chris1234
```

(my original example should be modified to include a `=` at the end of the look-ahead)

When you encounter a problem, I would only move onto a workaround (at the cost of a feature, in this case, allowing non-alphanumeric characters in a password) when you have figured out what the problem is and know for sure you can't fix it yourself. Otherwise I would've gone with your original solution (`awk ... | cut ...`).

Moreover, this behavior should be documented. If you want to restrict the keyspace, then you need to write in the README that `SWIFT_USERNAME` must be alphanumeric characters only. You must also describe how whitespace is not tolerated and the string should not be wrapped in quotation marks.

This is another advantage of

```bash
source swift/credentials.env
export SWIFT_USERNAME
export SWIFT_KEY
```

You could simply document that the format of `.env` (which is not a formally defined file format) obeys the rules of bash, which might be convenient because all the following are permitted:

```bash
SWIFT_USERNAME='$::'
SWIFT_USERNAME="$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c "${1:-60}")"
SWIFT_USERNAME="alice:bob"
SWIFT_USERNAME=alice:bob
SWIFT_USERNAME=$USER:chris1234
```

zrthxn replied:
@jennydaman 

I did try my best to understand what the problem was but with every combination I tried in the regex, I'd either just get a blank output or just a single character.

> What is the output of grep --version?

![Screenshot (70)](https://user-images.githubusercontent.com/35369637/118097547-761bfa00-b3f0-11eb-9ed9-e6bc670b4ebc.png)


> Have you tried using a container?

I have now, and it seems to work.
![Screenshot (71)](https://user-images.githubusercontent.com/35369637/118098563-ad3edb00-b3f1-11eb-95b0-3600fc973ca7.png)

So since exporting the credentials file gives us more flexibility, do you want me to revert to that? Or do you want me to continue using the grep method?

jennydaman replied:
My opinion is that `source ... && export SWIFT_USERNAME SWIFT_KEY` is more convenient (in this case, in high-security environments we would want to avoid using `source`).

If you'd like to continue troubleshooting your `grep` try using a different shell, 

```shell
sh -c 'grep -m 1 -Po "(?<=SWIFT_USERNAME).+" swift/credentials.env'
```

zrthxn committed:
Use PCRE look-behind and heredoc

jennydaman commented on files/startmain.sh:
```diff
@@ -57,14 +60,45 @@ if [ ! -z "${SWIFT_STORAGE_URL_SCHEME}" ]; then
+sed -i -e "s/user_userd_passd = keydd .admin/user_${echo $SWIFT_USERNAME | sed 's/:/_/g'} = ${SWIFT_KEY} .admin/g" /etc/swift/proxy-server.conf
+
+echo "Setting credentials in /etc/swift/dispersion.conf..."
+cat > /etc/swift/dispersion.conf << EOF
```
If we're using heredoc, `swift/credentials.env` and `swift/dispersion.conf` are not necessary and should be removed from this repo.

zrthxn replied:
Could remove `dispersion.conf` although maybe `credentials.env` would be helpful as a reference/guide?

jennydaman commented on files/startmain.sh:
```diff
@@ -57,14 +60,45 @@ if [ ! -z "${SWIFT_STORAGE_URL_SCHEME}" ]; then
+	echo "Using environment credentials..."
+elif [ -e /etc/swift/credentials.env ]; then
+	echo "Using default credentials file: /etc/swift/credentials.env ..."
+	SWIFT_USERNAME = grep -m 1 -Po '(?<=SWIFT_USERNAME.)[A-Za-z0-9\:]*' /etc/swift/credentials.env
```
Why not just `.+`? The `SWIFT_USERNAME` should be allowed to have non-alphanumeric characters like `-+=`

zrthxn replied:
For some reason `.+` seems [not to work](https://github.com/FNNDSC/docker-swift-onlyone/pull/2#discussion_r628088629)

jennydaman commented on files/startmain.sh:
```diff
@@ -57,14 +60,45 @@ if [ ! -z "${SWIFT_STORAGE_URL_SCHEME}" ]; then
+fi	
+
+if [ ! -z "${SWIFT_SET_RANDOMKEY}" ]; then
+	SWIFT_KEY=`pwgen 12 1`
```
I think random password generation is feature creep (UNIX philosophy is to do one thing and one thing well, let `pwgen` be something used outside of `docker-swift-onlyone`)

Alternatively, keep the random password feature for the sake of developer's convenience. But I'd prefer to use standard tools so we can drop the `pwgen` dependency.


```bash
generate_password ()
{
  head /dev/urandom | tr -dc A-Za-z0-9 | head -c "${1:-60}"
}
```

jennydaman replied:
Nevermind, developer's convenience would be a default password e.g. `chris:chris1234`

I'd suggest removing this feature and removing `pwgen`

zrthxn replied:
I just kept it because it was already there.
https://github.com/FNNDSC/docker-swift-onlyone/blob/224ae1f938d6d6bc1316ffed04855632818c3ff0/files/startmain.sh#L62

jennydaman replied:
Yeah I saw that, but I think it is a bad idea.

zrthxn committed:
Remove random key gen

zrthxn committed:
Remove dispersion.conf and credentials.env

zrthxn committed:
Revert to source-ing

zrthxn commented:
@jennydaman Reverted to source-ing

jennydaman commented:
@zrthxn my previous suggestion was to export the variables after sourcing, and avoid using `sed`

```bash
source $CREDENTIALS
export SWIFT_USERNAME SWIFT_KEY
```

It would also be good to write a few unit tests and use Github Actions. lmk if you need help with Github Actions.

zrthxn commented:
Oh alright. I do know about Actions, although I'm not sure what the unit tests would test for? Should they check the variables that the script uses or check the supervisord things?

zrthxn committed:
Export after source, avoid sed

jennydaman commented:
Just to test that a minimal setup works.

Basically, just an upload and download.

```bash
# idk if these commands actually work, I wrote them w/o testing them
swift -U chris:chris1234 -A http://localhost:8080/auth/v1.0  -K testing upload users LICENSE.txt --object-name apath/LICENSE.txt
swift -U chris:chris1234 -A http://localhost:8080/auth/v1.0  -K testing download users /tmp --prefix $FILE apath/LICENSE.txt
diff -q LICENSE.txt /tmp/LICENSE.txt
```

And then duplicate this test to make sure a custom password works as well, a docker secret works, ...

zrthxn committed:
Build image CI

Test image

Dependency

Test credentials

Put into same job

Dont checkout at every step

Install Swift

Try sudo?

apt update

Pip install

Pip install

zrthxn committed:
Heredoc for proxy-server.conf

Fix bug

zrthxn committed:
Line fix

zrthxn committed:
Upload and download test

zrthxn commented:
@jennydaman 
Tests are passing now
https://github.com/zrthxn/docker-swift-onlyone/runs/2610924992

If these are okay then I'll just change `.github/workflows/docker-image.yml` to run only on Pull requests to `master`

jennydaman commented on .github/workflows/docker-deploy.yml:
```diff
@@ -0,0 +1,25 @@
+    steps:
+    - uses: actions/checkout@v2
+    - name: Deploy to DockerHub
+      run: echo DockerHub
```
We do not need this file.

jennydaman commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,44 @@
+      run: |
+        docker build . --file Dockerfile --tag fnndsc/docker-swift-onlyone:test
+        sudo apt-get update
+        sudo apt-get install -y python3-pip
```
We do not need `sudo apt-get`. [`pip` is pre-installed](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md). For advanced Python config (which shouldn't be necessary) use [`actions/seup-python@v2`](https://github.com/actions/setup-python)

Here is an example for both how to use `pip` in vanilla Github Actions or `actions/setup-python@v2`
https://github.com/FNNDSC/chrisapp/blob/dcc6e4493b1592f2dff09e316847fe7e02e7dd42/.github/workflows/test.yml#L33-L34

jennydaman commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,44 @@
+        sudo apt-get install -y python3-pip
+        sudo pip install python-swiftclient
+
+    - name: Test OpenStack Swift [default]
```
Completely optional challenge, do you want to try parallelizing these tests using jobs?

zrthxn replied:
As far as I know (and have googled), I don't think that is possible in this particular case. I did try this before putting them into the same job, you can see that in the commit history.

Of course the three test steps can be run as parallel jobs, and they can use the `needs` keyword to wait for the build job to complete before starting. BUT since they need the docker image that was built by the previous job, they need to be on the same "runner", and at the moment there's no way to specify where a specific job (or step of a job) should run.

One solution could be to maybe upload the built docker image somewhere and then pull it into the test runners. But that'd just create more work than it's worth, with not enough speed improvements.

zrthxn replied:
So I was able to make it work with artifacts!

jennydaman commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,44 @@
+      run: |
+        docker run -d --name swift-onlyone -p 12345:8080 -v swift_storage:/srv -t fnndsc/docker-swift-onlyone:test
+        swift -A http://localhost:12345/auth/v1.0 -U chris:chris1234 -K testing upload users README.md --object-name ./README.md
+        swift -A http://localhost:12345/auth/v1.0 -U chris:chris1234 -K testing download users README.md --output /tmp/README.md 
```
Pedantic, I'd use `mktemp` here.

jennydaman commented on files/startmain.sh:
```diff
@@ -57,14 +60,47 @@ if [ ! -z "${SWIFT_STORAGE_URL_SCHEME}" ]; then
+storage_url_scheme = default
+use = egg:swift#tempauth
+user_admin_admin = admin .admin .reseller_admin
+user_$(echo $SWIFT_USERNAME | sed 's/:/_/g') = $SWIFT_KEY .admin" >> /etc/swift/proxy-server.conf
```
This solution forbids the username and password from containing the `:` character, would you

1. explicitly document this in the `README.md`?
2. change the `sed` expression to not use the `g` (as in global) flag so that the restriction applies to only the username, i.e. (potentially*) permit `:` as a valid password character

```bash
echo hello:world:goodbye | sed 's/:/_/'
hello_world:goodbye
```


*footnote: I don't know how this would actually end up working in `proxy-server.conf`

zrthxn replied:
The colon character cannot be allowed in the variable because although it'd be replaced by an underscore in `proxy-server.conf`, it'd probably cause issues in `dispersion.conf`. I'll document this in the readme.

zrthxn committed:
Requested changes done

zrthxn committed:
Test parallelize

Fix stupid mistake

jennydaman commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,107 @@
+      run: |
+        cd /tmp/artifacts
+        docker load < docker-image.tar
+        cd
```
This `cd` (back to the home dir) is unnecessary.
How about simply `docker load < /tmp/artifacts/docker-image.tar`?

jennydaman commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,107 @@
+
+    - name: Test OpenStack Swift [default]
+      run: |
+        sudo pip install python-swiftclient
```
Avoid `sudo` if unnecessary.
https://github.com/FNNDSC/docker-swift-onlyone/pull/2#discussion_r636179165

zrthxn replied:
I did try it without sudo, it seems to give an error.

jennydaman replied:
Sorry for missing this comment. 

Would you be specific about what error it gave you?

zrthxn replied:
I've removed sudo now, it works fine. sorry about the trouble.

jennydaman commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,107 @@
+    - name: Test OpenStack Swift [default]
+      run: |
+        sudo pip install python-swiftclient
+        docker run -d --name swift-onlyone -p 12345:8080 -v swift_storage:/srv -t fnndsc/docker-swift-onlyone:test
```
What is the volume for?

zrthxn replied:
Oh that was just the run command that was in the readme which got copied. Sorry about that.

jennydaman commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,107 @@
+      run: |
+        sudo pip install python-swiftclient
+        docker run -d --name swift-onlyone -p 12345:8080 -v swift_storage:/srv -t fnndsc/docker-swift-onlyone:test
+        swift -A http://localhost:12345/auth/v1.0 -U chris:chris1234 -K testing stat
```
I would put this in another step (for friendlier output).

zrthxn replied:
Sure. Will do.

jennydaman commented on .github/workflows/docker-image.yml:
```diff
@@ -0,0 +1,107 @@
+
+    - name: Test OpenStack Swift [env credentials]
+      run: |
+        sudo pip install python-swiftclient
```
Same advice from above applies here.

zrthxn committed:
Requested changes, separate into steps

zrthxn committed:
Dont use sudo on pip install

zrthxn commented:
@jennydaman 
That error was when I was using apt to install python-swiftclient.
With pip there was no need, I misremembered getting an error with pip.

https://github.com/zrthxn/docker-swift-onlyone/blob/e8dcf896ecc68beb421425ec69098912a5a446e1/.github/workflows/docker-image.yml#L16-L18

```
E: Could not open lock file /var/lib/dpkg/lock-frontend - open (13: Permission denied)
E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), are you root?
Error: Process completed with exit code 100.
```
Ref: https://github.com/zrthxn/docker-swift-onlyone/runs/2609372453?check_suite_focus=true#step:3:1124

jennydaman commented:
@zrthxn If `SWIFT_USERNAME` is defined in the environment, then `SWIFT_KEY` must also be defined by the environment.
i.e., it is unsupported to have a custom `SWIFT_USERNAME` and the default value for `SWIFT_KEY`.
Would you support this behavior, or document it in the README?

zrthxn commented:
Since both of them are important to the security of the storage, they should both be defined by the user any time that this container is started. So I'll document that the user must provide both of them. 
If they don't and they try to use just their custom username+password to access Swift, they'll get access denied which will hopefully tell them that something was wrong with their config.

zrthxn committed:
Update README.md
